File: backend/pipeline/change_detector.py
class ChangeDetector:
    def __init__(self, position_bucket_size=3):
        self.last_spoken_state = {}
        self.position_bucket_size = position_bucket_size

    def _bucket_position(self, bbox, frame_width):
        cx = (bbox[0] + bbox[2]) / 2
        bucket_width = frame_width / self.position_bucket_size
        idx = int(cx // bucket_width)
        labels = ["left", "center", "right"]
        return labels[min(idx, len(labels) - 1)]

    def _bucket_distance(self, area, frame_area):
        ratio = area / frame_area
        return "near" if ratio > 0.15 else "medium" if ratio > 0.04 else "far"

    def check(self, info_objects, frame_width, frame_height):
        frame_area = frame_width * frame_height
        changes, current_state = [], {}
        for obj in info_objects:
            cls = obj["class"]
            position = self._bucket_position(obj["bbox"], frame_width)
            distance = self._bucket_distance(obj["area"], frame_area)
            current_state[cls] = {"position": position, "distance": distance}
            prev = self.last_spoken_state.get(cls)
            if prev is None:
                changes.append({**obj, "position": position, "distance": distance})
            elif prev["position"] != position or prev["distance"] != distance:
                changes.append({**obj, "position": position, "distance": distance})
        self.last_spoken_state = current_state
        return changes

File: app/lib/pipeline/change_detector.dart
class ChangeDetector {
  static const int positionBucketSize = 3;
  Map<String, Map<String, String>> _lastSpokenState = {};

  String _bucketPosition(List<double> bbox, int frameWidth) {
    final cx = (bbox[0] + bbox[2]) / 2;
    final bucketWidth = frameWidth / positionBucketSize;
    final idx = (cx / bucketWidth).floor();
    const labels = ['left', 'center', 'right'];
    return labels[idx.clamp(0, labels.length - 1)];
  }

  String _bucketDistance(double area, int frameArea) {
    final ratio = area / frameArea;
    if (ratio > 0.15) return 'near';
    if (ratio > 0.04) return 'medium';
    return 'far';
  }

  List<Map<String, dynamic>> check(
    List<dynamic> infoObjects,
    int frameWidth,
    int frameHeight,
  ) {
    final frameArea = frameWidth * frameHeight;
    final changes = <Map<String, dynamic>>[];
    final currentState = <String, Map<String, String>>{};

    for (final obj in infoObjects) {
      final cls = obj.cls as String;
      final position = _bucketPosition(obj.bbox as List<double>, frameWidth);
      final distance = _bucketDistance(
          (obj.area as double).toDouble(), frameArea);

      currentState[cls] = {'position': position, 'distance': distance};

      final prev = _lastSpokenState[cls];
      if (prev == null) {
        changes.add({
          'cls': cls,
          'position': position,
          'distance': distance,
          'bbox': obj.bbox,
          'urgency': 'INFO',
          'motion': obj.motion,
          'area': obj.area,
        });
      } else if (prev['position'] != position ||
          prev['distance'] != distance) {
        changes.add({
          'cls': cls,
          'position': position,
          'distance': distance,
          'bbox': obj.bbox,
          'urgency': 'INFO',
          'motion': obj.motion,
          'area': obj.area,
        });
      }
    }

    _lastSpokenState = currentState;
    return changes;
  }

  void reset() {
    _lastSpokenState = {};
  }
}
